File: rb/110.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  recur(root)[0]
end

def recur(root)
  return [true, 0] if !root

  is_balanced, h_left = recur(root.left)
  return [false, nil] if !is_balanced

  is_balanced, h_right = recur(root.right)
  return [false, nil] if !is_balanced

  is_balanced = (h_right - h_left).abs <= 1
  return [is_balanced, [h_left, h_right].max + 1]
end



# raise error
def assert(tree)
  return 0 if not tree

  left, right = assert(tree.left), assert(tree.right)
  (left - right).abs > 1 ? raise : 1 + [left, right].max
end

def is_balanced(root)
  assert(root) && true rescue false
end

File: py/110.py
# DFS
class Solution(object):
    def isBalanced(self, root):
        def check(root):
            if not root:
                return 0

            l_depth, r_depth = check(root.left), check(root.right)

            if abs(l_depth - r_depth) > 1:
                raise

            return max(l_depth, r_depth) + 1

        try:
            check(root)
            return True
        except:
            return False



